modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java

```diff
@@ -213,7 +218,11 @@ private PageMemoryHashIndexStorage createOrRestoreHashIndex(IndexMeta indexMeta)
 
     /** {@inheritDoc} */
     @Override
-    public @Nullable BinaryRow read(RowId rowId, HybridTimestamp timestamp) throws StorageException {
+    public @Nullable ReadResult read(RowId rowId, HybridTimestamp timestamp) throws StorageException {
```

Comment:
btw, is it ok to return nulls now? I mean, what if you have a tombstone? There would be two ways to return empty value then, not cool. `ReadResult` itself should never be null.

Reply:
Tombstone just returns null. Are you sure we should wrap null values? Seems a bit like overkill, it is an allocation after all. Basically ReadResult can be null, but binary row inside it can't be null.

Reply:
Hm, probably a good idea will be to return something like ReadResult.EMPTY

Reply:
> Basically ReadResult can be null, but binary row inside it can't be null.

This is a faulty logic. You should write such tests. Imagine if you already have a value and you remove it in transaction. This means returning a write intent with null value and non-null timestamp. Is that correct? I think it is.